### PR TITLE
[WIP][SPARK-37798] PySpark Pandas API: Cross and conditional merging

### PR DESCRIPTION
**Pull request by ghost**
JIRA: https://issues.apache.org/jira/browse/SPARK-37798

Pandas currently supports a `how="cross"` merge which provides a cartesian product of the left/right tables. This can be achieved by doing a `spark.sql.dataframe.join(..., on=None, how="inner")`.

Additionally, I am currently in the middle of adding conditional merging in the pandas API (see PR here: https://github.com/pandas-dev/pandas/pull/42964). This is much easier to achieve in spark, since the functionality is already available, and we can trivially expose it in the pyspark pandas API. Due to the demand  of this functionality (countless SO/pandas issues either asking how to do this, or asking questions that would be solved by this), I think that this would be worth adding even before it makes it into the core pandas API.

These changes will be purely incremental on top of the existing API, and will be completely backwards compatible.

Still need to add tests and docstring examples.

**Examples:**

**Example DFs:**
```
>>> df1 = pd.DataFrame([['Bill', 23], ['Mary', 33], ['Ted', 36]], columns=['name', 'age'])
>>> df2 = pd.DataFrame([['President', 35], ['Senator', 30]], columns=['job', 'min_age'])
>>> df1
   name  age
0  Bill   23
1  Mary   33
2   Ted   36

>>> df2
         job  min_age
0  President       35
1    Senator       30
```

**Cross  Merge Example:**
```
>>> df1.merge(df2, how="cross")
   name  age        job  min_age
0  Bill   23  President       35
1  Bill   23    Senator       30
2  Mary   33  President       35
3  Mary   33    Senator       30
4   Ted   36  President       35
5   Ted   36    Senator       30
```

**Conditional Merge Example:**
```
>>> df1.merge(df2, on=lambda left, right: left.age >= right.min_age)
   name  age        job  min_age
0  Mary   33    Senator       30
1   Ted   36  President       35
2   Ted   36    Senator       30
```

### CODE REVIEW

File: python/pyspark/pandas/frame.py [hunk 1]
```diff
@@ -7648,52 +7654,6 @@ def to_list(os: Optional[Union[Name, List[Name]]]) -> List[Label]:
             else:
                 return [o if is_name_like_tuple(o) else (o,) for o in os]
 
-        if isinstance(right, ps.Series):
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/frame.py:7651]: Can we minimize the codes moving around? It's difficult to review. You can make a separate PR to refactor this first.

Comment (aa1371) [reply python/pyspark/pandas/frame.py:7651]: Sure I can do this. FYI, unfortunately, as part of this refactor implementation a large chunk of code was required to be indented one level, which will still cause a large diff block, even though the changes are fairly minimal.

How would you like me to structure the PRs? Would you rather:
(1) create a isolated code repositioning branch that will be merged to master in one PR, and then a second join-feature update branch that will be separately merged to master in a second PR? 
Or (2) create a code repositioning branch, with a PR to master, and then another join-feature update branch with a PR to the code repositioning branch. The join feature branch will get merged into the other branch first, then that branch to master.

I think (2) makes more sense, since if we don't agree on the implementation of the join feature, then the code repositioning will just be unnecessarily merged in noise.

Comment (HyukjinKwon) [reply python/pyspark/pandas/frame.py:7651]: Actually, I would prefer to create an isolated PR that refactors this into, for example, a separate private function or nested function as an example.

I think it's fine even without merging the join feature - this logic you are moving seems better be separate and factored out in any event.

Comment (aa1371) [reply python/pyspark/pandas/frame.py:7651]: @HyukjinKwon - I just pushed an update to address cleaning up the logic in the merge function (the first half of the function, up until the creation of the `joined_table`). And actually, I hope you don't mind, but I've also included the cross and conditional merge logic into the PR as well, because it actually is only a minimal marginal addition to the cleanup refactor to implement both features.

If we only want the PR to address cleaning up the function (i.e. no behavior change/new features), the only  change we would need to make to the PR is to reduce this block of code:
https://github.com/apache/spark/blob/f052ff0c90de487342a2da71a56cdee2b1b3099f/python/pyspark/pandas/frame.py#L7720-L7737

To this subset of that block of code:
```
left_key_names, right_key_names = self._resolve_merge_key_names(
    right, on, left_on, right_on, left_index, right_index
)
join_condition = _resolve_merge_join_condition_from_key_names(
    left_table, right_table, left_key_names, right_key_names
)
```

I think there will still be a bit to digest in the PR, but it's all related to the null-op cleanup refactor, after that's been processed the new feature additions are very simple to follow.

I think this would be a potentially good way to move forward, let me know your thoughts on this approach.

File: python/pyspark/pandas/frame.py [hunk 2]
```diff
@@ -7637,102 +7706,35 @@ def merge(
...
-                return [cast(Label, os)]
-            elif is_name_like_value(os):
-                return [(os,)]
-            else:
-                return [o if is_name_like_tuple(o) else (o,) for o in os]
-
         if isinstance(right, ps.Series):
             right = right.to_frame()
 
-        if on:
```
Comment (aa1371) [inline python/pyspark/pandas/frame.py:7654]: The logic from 7654-7693 has been moved verbatim to `self._resolve_merge_key_names`

File: python/pyspark/pandas/frame.py [hunk 3]
```diff
@@ -7637,102 +7706,35 @@ def merge(
...
+            if callable(on):
+                join_condition = on(left_table, right_table)
+            else:
+                # join condition based on column names
+                left_key_names, right_key_names = self._resolve_merge_key_names(
+                    right, on, left_on, right_on, left_index, right_index
+                )
+                join_condition = _resolve_merge_join_condition_from_key_names(
+                    left_table, right_table, left_key_names, right_key_names
+                )
```
Comment (aa1371) [inline python/pyspark/pandas/frame.py:7737]: If we keep only this highlighted portion of the block from 7720-7737, then this entire PR serves only as a code reorg, with no behavior or feature changes.

File: python/pyspark/pandas/frame.py [hunk 4]
```diff
@@ -7637,102 +7706,35 @@ def merge(
         As described in #263, joining string columns currently returns None for missing values
             instead of NaN.
         """
-
-        def to_list(os: Optional[Union[Name, List[Name]]]) -> List[Label]:
```
Comment (aa1371) [inline python/pyspark/pandas/frame.py:7641]: moved to global scope to declutter this function

File: python/pyspark/pandas/frame.py [hunk 5]
```diff
@@ -7637,102 +7706,35 @@ def merge(
...
-            if len(left_key_names) != len(right_key_names):
-                raise ValueError("len(left_keys) must equal len(right_keys)")
-
-        # We should distinguish the name to avoid ambiguous column name after merging.
-        right_prefix = "__right_"
-        right_key_names = [right_prefix + right_key_name for right_key_name in right_key_names]
-
-        how = validate_how(how)
-
-        def resolve(internal: InternalFrame, side: str) -> InternalFrame:
```
Comment (aa1371) [inline python/pyspark/pandas/frame.py:7697]: moved to global scope to declutter this function

File: python/pyspark/pandas/frame.py [hunk 6]
```diff
@@ -7637,102 +7706,35 @@ def merge(
...
-                        "left_on=None, right_on=None, left_index=False, right_index=False"
-                    )
-                left_key_names = list(map(self._internal.spark_column_name_for, to_list(common)))
-                right_key_names = list(map(right._internal.spark_column_name_for, to_list(common)))
-            if len(left_key_names) != len(right_key_names):
-                raise ValueError("len(left_keys) must equal len(right_keys)")
-
-        # We should distinguish the name to avoid ambiguous column name after merging.
-        right_prefix = "__right_"
-        right_key_names = [right_prefix + right_key_name for right_key_name in right_key_names]
```
Comment (aa1371) [inline python/pyspark/pandas/frame.py:7693]: The logic from 7654-7693 has been moved verbatim to `self._resolve_merge_key_names`

File: python/pyspark/pandas/frame.py [hunk 7]
```diff
@@ -7637,102 +7706,35 @@ def merge(
...
         right_table = right_internal.spark_frame.alias("right_table")
 
-        left_key_columns = [scol_for(left_table, label) for label in left_key_names]
-        right_key_columns = [scol_for(right_table, label) for label in right_key_names]
+        how = validate_how(how)
 
-        join_condition = reduce(
-            lambda x, y: x & y,
-            [lkey == rkey for lkey, rkey in zip(left_key_columns, right_key_columns)],
-        )
```
Comment (aa1371) [inline python/pyspark/pandas/frame.py:7735]: This logic has been moved to `_resolve_merge_join_condition_from_key_names ` in the global scope.
(not including the `validate_how` call)

File: python/pyspark/pandas/frame.py (hunk 1 above)
Comment (aa1371) [reply python/pyspark/pandas/frame.py:7651]: @HyukjinKwon - I just pushed an update to address cleaning up the logic in the merge function (the first half of the function, up until the creation of the `joined_table`). And actually, I hope you don't mind, but I've also included the cross and conditional merge logic into the PR as well, because it actually is only a minimal marginal addition to the cleanup refactor to implement both features. (This is the reason I kept the changes in this PR rather than a new PR)

If we only want the PR to address cleaning up the function (i.e. no behavior change/new features), the only  change we would need to make to the PR is to reduce this block of code:
https://github.com/apache/spark/blob/f052ff0c90de487342a2da71a56cdee2b1b3099f/python/pyspark/pandas/frame.py#L7720-L7737

To this subset of that block of code above:
https://github.com/apache/spark/blob/f052ff0c90de487342a2da71a56cdee2b1b3099f/python/pyspark/pandas/frame.py#L7731-L7737

I think there will still be a bit to digest in the PR, but it's all related to the null-op cleanup refactor, after that's been processed the new feature additions (from the first snippet above) are very simple to follow. I've added some comments to explain the cleanup reorg part.

I think this would be a good way to move forward, let me know your thoughts on this approach.

**Comment by aa1371** [comment]
@HyukjinKwon - explanation of latest update here: https://github.com/apache/spark/pull/35083#discussion_r777194747

**Comment by HyukjinKwon** [comment]
@aa1371 mind enabling GitHub Actions in your forked repository? Apache Spark repo uses the resources of GitHub Actions in your forked repository (see also https://github.com/apache/spark/pull/35083/checks?check_run_id=4686215216).

**Comment by HyukjinKwon** [comment]
cc @itholic @ueshin @xinrong-databricks FYI

**Comment by aa1371** [comment]
@HyukjinKwon - workflows enabled. FYI - tests have not been added for these features yet. Will wait until there is buy-in on them first.

**Comment by aa1371** [comment]
added some preliminary tests, can add onto these later.

**Comment by aa1371** [comment]
@HyukjinKwon - thoughts on the PR in its current state?

Also, is it possible to see the specific formatting issues caught by the black formatter in the CI? It seems to contain some different black configuration than when I run it dev/format-python locally

**Comment by zero323** [comment]
> Also, is it possible to see the specific formatting issues caught by the black formatter in the CI?

Not at the moment. 

> It seems to contain some different black configuration than when I run it dev/format-python locally

This seems to be your current offender

```patch
diff --git a/python/pyspark/pandas/frame.py b/python/pyspark/pandas/frame.py
index 949e3e601b..da91da59bc 100644
--- a/python/pyspark/pandas/frame.py
+++ b/python/pyspark/pandas/frame.py
@@ -359,7 +359,8 @@ RIGHT_MERGE_PREFIX = MERGE_PREFIX_TEMPLATE.format("right")
 
 
 def _resolve_internal_merge_frame(
-    internal: InternalFrame, side: Optional[str] = None,
+    internal: InternalFrame,
+    side: Optional[str] = None,
 ) -> InternalFrame:
     internal = internal.resolved_copy
     if side is None:
```

In general you might want to resync this with master.
